### ISSUE TITLE
Semicolons.  Sorta

### ISSUE THREAD

**Issue by laicosly** [airbnb/javascript#394]
The argument presented for using semicolons shows one of the few areas that semicolons are actually needed (anonymous self-enclosing functions, and for i loops).

Semicolons everywhere is clutter, use them only where required.

https://github.com/airbnb/javascript#20.1

**Comment by justjake** [comment]
That's a perfectly reasonable stance on semicolons, and there are plenty of other styleguides out there that use that rule -- feross/standard comes to mind.

**Comment by goatslacker** [comment]
As a fellow member of the no semicolons in my JS camp I couldn't agree more. But this is just the guide that we follow at Airbnb. I don't expect to see this change anytime soon.

**Comment by goatslacker** [comment]
If you're interested in this guide sans semicolons then https://github.com/goatslacker/javascript

**Comment by laicosly** [comment]
;very cool;
i just thought i'd suggest it;
thank you for the style-guide recommendations;
i am digging the feross/standard one so far;
carry on;

**Comment by goatslacker** [comment]
Yeah two semicolons before and after each line is good practice, you know, for safety from the evil ASI monster that no one wants to bother to understand.

**Comment by laicosly** [comment]
;; @goatslacker  :+1:  ;;

**Comment by hshoff** [comment]
};

**Comment by seb0zz** [comment]
Can anyone tell me why not using semicolons everywhere is a bad practice? I don't really find a satisfying argumentation anywhere in this repository.

**Comment by ljharb** [comment]
@seb0zz that's a long argument, full of FUD on both sides, which really boils down to subjective preference. Personally I find the rules of ASI harder to remember and more likely to cause bugs, than the rules of semicolons - and ideologically I consider semicolons *so* required by the spec that it puts them in for you even if you forget :-) but it's really about what you and your team find consistently sensible.

**Comment by lencioni** [comment]
@seb0zz Here's an example that I recently saw via Twitter: https://tonicdev.com/tolmasky/comparison

**Comment by seb0zz** [comment]
@lencioni are you sure that the link is correct? Doesn't look like anything about semicolons ;)

**Comment by lencioni** [comment]
@seb0zz Yes, it is demonstrating some cases where not having semicolons introduces ambiguity which causes unexpected behavior. I think the point that the author is making is that while the ES5 example is pretty unlikely to happen in the real world, the ES6 example is likely to happen in the real world because we now have array destructuring.

TL;DR:

This code with semicolons will assign `num` to `1`:

```js
var options = [1, 2, 3, 4];
[num] = options;
```

whereas this code without a semicolon will give you a ReferenceError, and `options` and `num` will both be `undefined`:

```js
var options = [1, 2, 3, 4]
[num] = options;
```

**Comment by seb0zz** [comment]
Thanks @lencioni now I get the point. Array destructuring is a a real argument for semicolons. Even though I would not really expect different evaluations only because of the missing semicolons as we have a new line between :-1:

**Comment by ljharb** [comment]
(That common and incorrect expectation is one of the arguments in favor of semicolons)

**Comment by seb0zz** [comment]
@ljharb what makes that expectation "incorrect"?

**Comment by ljharb** [comment]
@seb0zz the expectation that newlines terminate statements in JS - they do not.

**Comment by seb0zz** [comment]
@ljharb and why can i write 

```
var a = 1
var b = 2
````
but not
```
var a = 1 var b = 2
```

**Comment by ljharb** [comment]
@seb0zz because the rules for ASI are generally found to be confusing, and that's the way JS works ¯\\\_(ツ)_/¯

**Comment by seb0zz** [comment]
@ljharb you are right ;) nevertheless everything would be easier if we wouldn't have semicolons and line breaks would terminate statements :tongue:

**Comment by softwarespot** [comment]
@seb0zz But then how would you minify if every line has to terminate with new line character? You're talking 1 or 2 bytes extra per line, which might not seem a lot at first, but then think of bandwidth that would be taken up for something that could be resolved by the use of semi-colons.

**Comment by ljharb** [comment]
you wouldn't have to - a newline character and a semicolon character are the same number of bytes.

**Comment by softwarespot** [comment]
On Windows it's 2 bytes if you're not paying attention and minifiers can be quite smart at removing semi-colons that aren't required.

**Comment by seb0zz** [comment]
@softwarespot the minifier could replace line breaks with semicolons where necessary

**Comment by softwarespot** [comment]
I was under the assumption @seb0zz was suggesting that line breaks would replace semi-colons, like a BASIC-like language. But I agree. Just wanted to join in the conversation

**Comment by seb0zz** [comment]
I would suggest that line breaks terminate statements and semicolons are optional and can also be used to terminate statements (for instance if you had multiple statements in one line). But this has nothing to do with this issue anymore ;)

**Comment by langri-sha** [comment]
@seb0zz then indentation might start to matter as well. Either way, this repository is a tool and then there are other tools out there. As time goes on, these tools tend to get better. In recent years, they've been getting really good really fast.

Historically, there were no linters. Then they came, but I couldn't configure them. Afterwards, I could start configuring them, but then there was a problem with imposing certain things because the syntax would allow us all to introduce inconsistencies. Today I'm using `babel-eslint` and when I lint, a small :palm_tree: is inflated and I look at it and I unleash the lint :dragon:. By 2020, my editor is probably going to be writing code straight into a linter. By 2030 linters will be writing most of our code and I can start doing something else in my life. I'd like to grow :tomato:.

**Comment by hshoff** [comment]
Configure your linter for your team 👥 not for individuals 👤.
This makes the code look like 👤 even though it was written by 👥.

We open sourced ours so others could use it as a starting point for their own. http://nerds.airbnb.com/our-javascript-style-guide/

3.5 years later and a transition from ES5 to ES6, a lot has changed in the style guide but our team still finds a lot of value from it.

Thanks to everyone for their thoughts on the all-important semicolon! We’re going to continue to require it for the forseeable future, so we’ll be locking this thread. Please feel free to fork the guide and make the changes that are best for you and your team. Remember, having a style guide is more important than the rules within. ❤️ 

## };
